src/chunk.c

```diff
@@ -1845,17 +1968,29 @@ ts_chunk_exists_relid(Oid relid)
  * on the chunk after the raw data was dropped. Otherwise, we'd have dangling
  * chunk ids left over in the materialization table. Preserve the space dimension
  * info about these chunks too.
+ *
+ * When chunk rows are preserved, the rows need to be updated to set the
+ * 'dropped' flag to TRUE. But since this produces a new tuple into the
+ * metadata table we will process also the new tuple in the same loop, which
+ * is not only inefficent but could also lead to bugs. For now, we just ignore
+ * those tuples (the CHUNK_ALREADY_MARKED_DROPPED case), but ideally we
+ * shouldn't scan the updated tuples at all since it means double the number
+ * of tuples to process.
  */
-static void
+static ChunkDeleteResult
 chunk_tuple_delete(TupleInfo *ti, DropBehavior behavior, bool preserve_chunk_catalog_row)
 {
 	FormData_chunk form;
 	CatalogSecurityContext sec_ctx;
 	ChunkConstraints *ccs = ts_chunk_constraints_alloc(2, ti->mctx);
+	ChunkDeleteResult res;
 	int i;
 
 	chunk_formdata_fill(&form, ti->tuple, ti->desc);
 
+	if (preserve_chunk_catalog_row && form.dropped)
+		return CHUNK_ALREADY_MARKED_DROPPED;
```

Comment:
why is this not an error condition? How do we run into this case where we try to mark the same chunk as dropped more than once?

Reply:
The metadata table is scanned using the UNIQUE <schema, name> index, which doesn't include the dropped column. Therefore, there is no filter on dropped chunks and this acts as the filter instead. 

Also not that there is a case when we'd like to DELETE also chunk entries that are "dropped", so we cannot always filter rows with dropped=true.